src/main/scala/rps/Game.scala

```diff
@@ -2,69 +2,64 @@ package rps
-    val User: String = "1" // User wins
-    val Cpu: String = "2" // Cpu wins
-  }
-
   def play(): Unit = {
-    println(s"""Choose your weapon:
-    ${Weapon.Rock} - ${matchWeapon(Weapon.Rock)}
-    ${Weapon.Paper} - ${matchWeapon(Weapon.Paper)}
-    ${Weapon.Scissors} - ${matchWeapon(Weapon.Scissors)}
-    """)
-    val rawUserInput = readLine()
-    val userWeapon = rawUserInput
-    val cpuWeapon = generateComputerMove()
+    val menu = Move.moves
+      .map(m => s"${Move.encode(m)} - ${Move.print(m)}")
+      .mkString("\n")
+    println("Choose your weapon:")
+    println(menu)
 
-    val result = checkWinner(userWeapon, cpuWeapon)
+    val rawUserInput = readLine()
+    val userMove = Move.decode(rawUserInput)
+    val cpuMove = generateComputerMove()
 
-    println(s"You chose: ${matchWeapon(userWeapon)} | your opponent chose: ${matchWeapon(cpuWeapon)}")
-    println(announceWinner(result))
-  }
+    val printedUserMove =
+      userMove.map(Move.print).getOrElse("💩 An invalid weapon")
+    val printedCpuMove = Move.print(cpuMove)
+    println(
+      s"You chose: ${printedUserMove} | your opponent chose: ${printedCpuMove}"
+    )
 
-  def matchWeapon(weapon: String): String = {
-    weapon match {
-      case Weapon.Rock     => "🪨 Rock"
-      case Weapon.Paper    => "📄 Paper"
-      case Weapon.Scissors => "✂️ Scissors"
-      case _               => "💩 An invalid weapon"
-    }
+    val result = checkWinner(userMove, cpuMove)
+    println(announceResult(result))
   }
 
-  def announceWinner(result: String): String = {
+  def announceResult(result: GameResult): String = {
     result match {
-      case Winner.Draw => "It's a draw ✍️"
-      case Winner.User => "You win 🎉"
-      case _           => "You lose 🤷"
+      case GameResult.Draw     => "It's a draw ✍️"
+      case GameResult.UserWins => "You win 🎉"
+      case GameResult.CpuWins  => "You lose 🤷"
+      case GameResult.DumbUser =>
+        "You lose 🤷 (next time choose a valid weapon)"
     }
   }
 
-  def checkWinner(userMove: String, computerMove: String): String = {
+  def checkWinner(
+      userMove: Option[Move],
+      computerMove: Move
+  ): GameResult = {
     (userMove, computerMove) match {
-      case (x, y) if x == y              => Winner.Draw
-      case (Weapon.Rock, Weapon.Scissors)  => Winner.User
-      case (Weapon.Paper, Weapon.Rock)     => Winner.User
-      case (Weapon.Scissors, Weapon.Paper) => Winner.User
-      case _                               => Winner.Cpu
-      // default case also handles an invalid weapon (e.g. user inputs "4")
+      case (Some(x), y) =>
+        (x, y) match {
+          case (x, y) if x == y            => GameResult.Draw
+          case (Move.Rock, Move.Scissors)  => GameResult.UserWins
+          case (Move.Paper, Move.Rock)     => GameResult.UserWins
+          case (Move.Scissors, Move.Paper) => GameResult.UserWins
+          case _                           => GameResult.CpuWins
```

Comment:
qui secondo me c'è uno spunto interessante su come usare il pattern matching

in questo caso secondo me va bene così perché nessuno ragionevolmente aggiungerà mai nuove mosse al gioco

in un caso in cui realisticamente potrebbero comparire nuovi valori invece penso che varrebbe la pena indicare tutto esplicitamente perché funzionino i controlli di esaustività, quindi così, anche se viene meno compatto:
```scala
case (Rock, Rock) | (Paper, Paper) | (Scissors, Scissors) => GameResult.Draw
case (Rock, Scissors) | (Paper, Rock) | (Scissors, Paper) => GameResult.UserWins
case (Rock, Paper) | (Paper, Scissors) | (Scissors, Rock) => GameResult.CpuWins
```

Reply:
Sì, sono d'accordo che i benefici dell'exhaustiveness check sono molto apprezzabili.

Just for fun: ho ipotizzato l'uscita di _RPS 2: The Lizard Spock Expansion_ per il prossimo training (non odiarmi XD), è solo una mia impressione oppure è abbastanza illeggibile? 

```scala
case (Rock, Rock) | (Paper, Paper) | (Scissors, Scissors) | (Lizard, Lizard) | (Spock, Spock)                                                                                               => GameResult.Draw
case (Rock, Scissors) | (Paper, Rock) | (Scissors, Paper) | (Rock,Lizard) | (Paper,Spock) | (Scissors,Lizard) | (Lizard,Spock) | (Lizard,Paper) | (Spock, Scissors) | (Spock, Rock)           => GameResult.UserWins
case (Rock, Paper) | (Paper, Scissors) | (Scissors, Rock) | (Rock, Spock) | (Paper, Lizard) | (Scissors, Spock) | (Lizard, Scissors) | (Lizard, Rock) | (Spock, Paper) | (Spock, Lizard)    => GameResult.CpuWins

// Formattato

case (Rock, Rock) | (Paper, Paper) | (Scissors, Scissors) |
              (Lizard, Lizard) | (Spock, Spock) =>
            GameResult.Draw
          case (Rock, Scissors) | (Paper, Rock) | (Scissors, Paper) |
              (Rock, Lizard) | (Paper, Spock) | (Scissors, Lizard) |
              (Lizard, Spock) | (Lizard, Paper) | (Spock, Scissors) |
              (Spock, Rock) =>
            GameResult.UserWins
          case (Rock, Paper) | (Paper, Scissors) | (Scissors, Rock) |
              (Rock, Spock) | (Paper, Lizard) | (Scissors, Spock) |
              (Lizard, Scissors) | (Lizard, Rock) | (Spock, Paper) |
              (Spock, Lizard) =>
            GameResult.CpuWins
```

È un po'un caso limite, visto che è un gioco in cui i possibili stati aumentano esponenzialmente ad ogni nuova mossa aggiunta. Ma in casi come questi abbiamo frecce migliori al nostro arco? O io dev mi rassegno ad impazzire?

Reply:
Si può migliorare un pochino annidando i pattern all'interno delle tuple, per esempio scrivendo
```
(Rock, Scissors | Lizard)
```
invece di
```
(Rock, Scissors) | (Rock, Lizard)
```

In questo mi sa che impazzisci lo stesso. Però in altri casi scrivendo così potresti riuscire a compattare abbastanza e rendere più esplicite cose tipo, per esempio se stai scrivendo un reducer, "in questo stato solo quest'azione fa qualcosa e le altre non fanno niente". Puoi riuscire lo stesso a elencare tutte le altre ma con `(Stato, Azione1 | ... | AzioneN)` che magari non è troppo tremendo.

Abbiamo una cosa del genere [qui](https://github.com/buildo/cash-flow-assistant/blob/master/api/package_/src/main/scala/cashflowassistant/package_/ProjectPackageService.scala#L519).